### PR TITLE
Unify data loading interfaces

### PR DESCRIPTION
**Pull request by rhaps0dy**
Add more metrics

### CODE REVIEW

File: acdc/greaterthan/utils.py [hunk 1]
```diff
@@ -106,7 +125,55 @@ def get_year_data(num_examples, model):
...
+        "greaterthan": partial(greaterthan_metric, tokens=test_data),
+        "kl_div": partial(
+            kl_divergence,
+            base_model_logprobs=base_test_logprobs,
+            mask_repeat_candidates=None,
+            last_seq_element_only=True,
+        ),
+    }
+
+    return AllDocstringThings(
```
Comment (rhaps0dy) [inline acdc/greaterthan/utils.py:168]: The class is called AllDocstringThings but it's generic, let me change it

File: subnetwork_probing/train.py [hunk 2]
```diff
@@ -106,7 +108,18 @@ def visualize_mask(model: HookedTransformer) -> tuple[int, list[TLACDCInterpNode
                 if mask_sample < 0.5:
                     nodes_to_mask.append(node)
 
-    assert len(mask_scores_for_names) == 3 * number_of_heads * number_of_layers
+        # MLP
+        node_name = f"blocks.{layer_index}.hook_mlp_out"
+        node_name_list.append(node_name)
+        node = TLACDCInterpNode(node_name, TorchIndex([None]), EdgeType.PLACEHOLDER)
```
Comment (rhaps0dy) [inline subnetwork_probing/train.py:114]: Is this the right node to remove when we're masking out an MLP? @ArthurConmy

File: acdc/main.py [hunk 3]
```diff
@@ -281,18 +266,25 @@
...
 
-exp.save_edges("another_final_edges.pkl") 
+exp.save_edges("another_final_edges.pkl")
+
+
+# Save test dataset
+
+with torch.no_grad():
+    exp.ref_ds = things.test_patch_data
+    exp.model.reset_hooks()
```
Comment (ArthurConmy) [inline acdc/main.py:279]: I don't think you want to reset hooks and setup the second cache. This will leave the model with no hooks on it at all, so `exp.model(things.test_data)` will be the same as `tl_model(things.test_data)` on a fresh `HookedTransformer`. I think you just want to delete these two lines

File: acdc/tracr/utils.py [hunk 4]
```diff
@@ -236,57 +238,115 @@ def get_perm(n, no_fp = True):
...
-        n = len(data_tens)
-        data_tens = data_tens.long()
-        patch_data_indices = get_perm(n)
+        patch_data_indices = get_perm(len(data_tens))
         warnings.warn("Test that this only considers the relevant part of the sequence...")
         patch_data_tens = data_tens[patch_data_indices]
         base_model_logprobs = F.log_softmax(tl_model(data_tens), dim=-1)
-        metric = partial(kl_divergence, base_model_logprobs=base_model_logprobs, mask_repeat_candidates=None)
+
+        if metric_name == "kl_div":
```
Comment (ArthurConmy) [inline acdc/tracr/utils.py:263]: Fairly certain this is updated in a later PR

**Review by ArthurConmy — CHANGES_REQUESTED** [review]
One bug (I think) and a few comments, then pls merge : )

File: acdc/main.py (hunk 3 above)
Comment (rhaps0dy) [reply acdc/main.py:279]: We talked about this, I thought re-running `.setup_second_cache()` was the way to run ACDC on a test set. It seems to output sensible things.

Also it internally calls `.reset_hooks()` so it's almost certainly fine to run it before. I can delete it because it's superfluous but I'm *not* going to delete `setup_second_cache` after `exp.ref_ds = things.test_patch_data` unless you make a better case.

File: acdc/tracr/utils.py (hunk 4 above)
Comment (rhaps0dy) [reply acdc/tracr/utils.py:263]: oh okay, I wasn't aware we wanted to do L2 for both tracr tasks. I can change it and re-run this task with l2.

File: acdc/main.py (hunk 3 above)
Comment (rhaps0dy) [reply acdc/main.py]: How's the new change @ArthurConmy ?
